src/parser/document.ts

```diff
@@ -34,7 +34,16 @@ export abstract class ParsedDocument<AstNode, Visitor> {
    */
   astNode: any;
 
-  private _locationOffset: LocationOffset|undefined;
+  sourceRange: SourceRange;
+
+  private readonly _locationOffset: LocationOffset|undefined;
+
+  /**
+   * The 0-based offsets into `contents` of all newline characters.
+   *
+   * Useful for converting between string offsets and SourcePositions.
+   */
+  readonly newlineIndexes: number[] = [];
```

Comment:
Will use this more in a followup PR. Will make some source range operations in ParsedHtmlDocument much faster.